bluesky/simulators.py

```diff
@@ -114,3 +120,424 @@ def check_limits(plan):
+    is a dictionary containing devices (detectors, axes) that have been
+    updated, as keywords and a value indicating the update status. For 'set'
+    the update status is the latest 'position' that it is to be set and for
+    'trigger' the update status is the number of times since the beginning of
+    the scan or since an 'unstage' event that the device has been triggered.
+"""
+# GROUP THE DIFFERENT COMMAND TYPES
+# The commands that start/end a 'group', but not flyer commands.
+_group_start_cmds = ['set', 'trigger', 'kickoff']
+_group_end_cmds = ['wait']
+# The commands that aren't  groupable but still have statistics.
+_timed_cmds = ['stage', 'unstage']
+# The commands that only require network communication.
+_communication_cmds = ['read', 'describe', 'open_run', 'close_run']
+# Define some times associated with network communications and internal
+# state tasks. I am not yet sure how to accurately estimate these times as
+# they are unlikely to be 'static' and measuring them for statistics will
+# take as much time as performing them, they are including incase a good
+# solution for this is found.
+_communication_time = _TimeStats(0, 0)  # An est_time/std_dev named tuple
+_internal_state_time = _TimeStats(0, 0)  # An est_time/std_dev named tuple
+
+
+def est_time(plan):
+    '''The generator function estimates the time to complete a plan with
+    a yield at every message.
+
+    This method takes in a plan and yields out a tuple with the structure
+    (msg, estimated stop time, estimated start time). Where msg is the
+    message from the plan and stop/start times are (est_time, std_dev)
+    named tuples which provides the end/start time for each message taking
+    the scan start as zero. Note that for 'groupable' actions (like 'set'
+    and 'trigger') the action msg will start and stop within the group time.
+
+    Parameters
+    ----------
+    plan : generator.
+        The bluesky plan that the est_time is to be estimated for.
+
+    Returns
+    -----------------
+    out_time : named tuple.
+        A named tuple containing 3 items, the message specifed in the plan,
+        the est_time/ std_dev named tuple for cumulative time and the est_time
+        /std_dev named tuple for the message time.
+    '''
+
+    # create a reference to a dictionary called plan_history to capture set
+    # changes during the plan for later modification.
+    plan_history = {'set': {}, 'trigger': {}}
+    # create a dictionary for storing the info regarding groups in progress.
+    groups = defaultdict(_TimeStats())
+
+    # used to hold the cumulative time for the plan
+    out_time = _TimeStats(0, 0)
+
+    for msg in plan:
+        # case 1: commands that are 'groupable'
+        if msg.command in _group_start_cmds:
+            # add set and trigger commands to plan_history
+            if msg.command == 'set':
+                plan_history['set'][msg.obj.name] = msg.args[0]
```

Comment:
Why pick off just `msg.args[0]` here? What about any others `args` and `msg.kwargs`?